### PR TITLE
Off canvas menu editing: adds list view to navigation block

### LINKED ISSUE #42257: Navigation List View: Introduce navigation editable tree view in the inspector controls
Building a custom navigation menu is fiddly. We have in various versions explored surfacing a List View interface for managing and building the contents of your navigation. Let's surface the list view in the inspector.

**When no menus exist**

In this state, all pages (Page List) is shown:

Navigation_ No menus

**A single menu exists**

In this state, this one menu is loaded by default:

Navigation_ One menu

**Multiple menus exist**

In this state, the context defines which menu is loaded by default. If the navigation block is inside the Header template part, "Header navigation" is loaded by default, even if "Navigation" and "Navigation 2" also exist. 

Navigation_ Multiple menus

To facilitate discoverability of this inspector interface, we should try to **enable the click overlay on the navigation block**, just like it is for template parts in the site editor. That way the first click on your navigation will always select the full block, and surface the inspector interface.

See also #40204 for an adjacent effort to divide the inspector into tabs, which will be useful for 30+ items in a navigation block.

---

This issue was updated Sep. 28.

<details>
<summary>See initial proposal</summary>

## What problem does this address?
Many users, both new and existing, have found the navigation block difficult to use. For new WordPress users, there are certain UI quirks that get in the way and jumping from parent to child blocks is challenging. For existing WordPress users who are new to FSE, there is a unique mix of directly editing the nav block (which is a new paradigm) and editing menus globally using the navigation sidebar.

image

## What is your proposed solution?

https://user-images.githubusercontent.com/1072756/177929562-bec50ae8-5cf1-47ff-93d6-f74ac86b33cb.mp4

Id like to propose that we simplify by leaning more heavily into editing menus locally/on canvas via the navigation block and emphasise template parts / re-usable blocks / synced patterns as the primary way of editing site elements globally. 

A quick way of moving in this direction is simply moving the menu management experience into the navigation block settings sidebar, and removing the existing global navigation sidebar. We'd also enable adding navigation menu items from this management experience which removes the need to jump in and out of child blocks. Adding menu items makes use of the existing add menu item dialog. Menu editing can still be done directly on canvas, in this case I think the redundancy is fine.

The video shown here makes use of some of the ideas found [here](https://github.com/WordPress/gutenberg/issues/41717), namely surfacing child layers within patterns (e.g. header) which may be a more accessible way of selecting the navigation block for low tech users.

On top of the above change, a few other optimisations to make:
- The add menu item dialog removes the ability to transform from the menu item selection stage and instead adds a back button
- As seen in the screenshot above, the navigation toolbar can block menu selection when in an empty state
- It also partially covers up the new block button as shown below

image

- This may just be a me issue but I've found the current state very buggy. Selecting the navigation block sometimes puts the editor into an unusable state, preventing me from selecting any other layer. Video attached.

https://user-images.githubusercontent.com/1072756/178391778-9a619e5a-2383-4e15-885e-78742bf30a16.mp4

I realise this contradicts from [this issue](https://github.com/WordPress/gutenberg/issues/36667) somewhat but if we introduce the idea of browsing/searching content in the sidebar then a dedicated UI for editing menu's becomes less important. We could still provide a way for managing menus outside the editor if there is value beyond just not having to enter the editor.

</summary>

### LINKED ISSUE #45444: Navigation List View: Move the navigation selector to the dot menu
As outlined in https://github.com/WordPress/gutenberg/issues/42257, we should move the navigation selector to the dot menu, so that it's not a primary action:

Screenshot 2022-11-01 at 11 36 10

Should be closed by https://github.com/WordPress/gutenberg/pull/44534

### PR DESCRIPTION
**Pull request by draganescu**
## What?

Part of #42257
Closes #45444

Adds a list view to the navigation block inspector.

## Why?

Using and locating nested items that are hidden in navigation submenus is difficult.

## How?

Copies the implementation in the site editor navigation sidebar which uses a block editor provider to allow the list view to not lose selection to the block in the canvas. 

## Testing Instructions

1. Switch to this PR
2. Go to Gutenberg > Experiments
3. Toggle on "Test a new off canvas editor for navigation block "
4. Save
5. Create a post
6. Add a navigation block
7. Create a block menu with sample items if one is not selected (it has to be a menu)
8. Select the block
9. Notice the new list view in the block inspector
10. Notice the dotted "more menu" in the panel that contains the list view, this is where the menu selector is now.

## Screenshots or screencast 

https://user-images.githubusercontent.com/107534/199258779-af083416-e0c1-4169-b3bd-233a4adfccd3.mp4

## Todo / known issues

- [ ] ~~Add an "Add menu item" item to the end of the list view (asked if this should be a list view item in general)~~
- [ ] Bring back the appender inside the nested branches of the list view
- [x] ~~Fix spacing of manage menus (asked for mockup in design issue)~~
- [x] Move Manage menus to the advanced area
- [x] Cleanup places where list view is not needed (e.g. when there is no menu)
- [x] Find other ways to disable canvas selection via Listview and drop the use of block editor provider.
- [ ] Use [a scrollbar](https://github.com/WordPress/gutenberg/issues/42257#issuecomment-1262499986) on the list view's container only for nested expansions, instead of overflowing the entire inspector.

### CODE REVIEW

File: packages/block-library/src/navigation/edit/index.js [hunk 1]
```diff
@@ -736,6 +758,16 @@ function Navigation( {
...
+						<BlockEditorProvider
+							value={ navigationInnerBlocks }
+							onChange={ onChange }
+							onInput={ onInput }
+						>
+							<ListView
+								blocks={ navigationInnerBlocks }
+								isExpanded={ true }
+							/>
+						</BlockEditorProvider>
```
Comment (talldan) [inline packages/block-library/src/navigation/edit/index.js]: Is it needed here when there are no inner blocks?

File: packages/block-library/src/navigation/edit/index.js [hunk 2]
```diff
@@ -655,6 +667,16 @@ function Navigation( {
...
```
Comment (talldan) [inline packages/block-library/src/navigation/edit/index.js]: In this case the inner blocks are uncontrolled, so I don't think `BlockEditorProvider` is needed (there isn't an entity to edit with `onChange` and `onInput`).

File: packages/block-library/src/navigation/edit/index.js [hunk 3]
```diff
@@ -213,6 +219,12 @@ function Navigation( {
 		hasResolvedCanUserCreateNavigationMenu,
 	} = useNavigationMenu( ref );
 
+	const [ navigationInnerBlocks, onInput, onChange ] = useEntityBlockEditor(
+		'postType',
+		'wp_navigation',
+		{ id: ref }
+	);
```
Comment (talldan) [inline packages/block-library/src/navigation/edit/index.js]: Is `useEntityBlockEditor` and the `BlockEditorProvider` definitely needed?

Normal list view works fine for editing navigation inner blocks, and doesn't have this.

If this is copied from the navigation sidebar, I believe it is required there because it's possible to show a menu that isn't within the editor content (the blocks aren't in the block editor's store). 

My understanding is that this will always show a menu that's also within the editor content, so I think it may not be needed.

Comment (draganescu) [reply packages/block-library/src/navigation/edit/index.js]: This _is_ a quick late in the day PR, should have made it a draft. However, without using a block entity provider, selecting an item from ListView selects the block in the canvas. 🤷🏻

Comment (talldan) [reply packages/block-library/src/navigation/edit/index.js]: > However, without using a block entity provider, selecting an item from ListView selects the block in the canvas. 🤷🏻

That's intentional, because List View is a way to select blocks in the canvas.

The reason why BlockEditorProvider would have this effect is because it essentially creates an entirely separate block editor store, with different selection state to the main block editor.

For this new List View in the nav block, what should selecting a block in List View do?

Comment (draganescu) [reply packages/block-library/src/navigation/edit/index.js]: Nothing, it should select the item _in the list view_ so users can drag/drop/use the more menu to uodate the navigation in a tree view.

**Comment by jasmussen** [comment]
Thank you for working on this! 🎉

Even in this state: click-through and list view management, it feels like the potential is there. Here's a GIF:

after

A few notes:

* Drag and drop doesn't appear to work entirely correctly. 
* Can "Manage menus" exist inside the menu selection dropdown?
* Selecting a menu item doesn't select it in the canvas. This is intentional as otherwise it would put focus there and the inspetor would be replaced. But maybe the selection style should be different, perhaps black instead of blue. Nothing to block this PR, just thinking out loud.
* Perhaps a question for @jameskoster — I can't recall the latest on clickthrough. When I select a template with clickthrough, there's a blue overlay. There isn't in this branch when I select the menu. Should there be?

I'm missing the "Create new item" button, which is going to be the next tricky step to solve. The basic idea is to let you build the entire menu, including sorting, moving into submenus, etc., in this panel alone. I imagine a simple way to start is that this would insert a Page Link with the dropdown open, this one:

Screenshot 2022-09-29 at 10 24 41

Would that be possible/feasible?

File: packages/block-library/src/navigation/edit/index.js (hunk 3 above)
Comment (talldan) [reply packages/block-library/src/navigation/edit/index.js]: > Nothing, it should select the item in the list view so users can drag/drop/use the more menu to uodate the navigation in a tree view.

Selecting in List View doesn't really mean anything though. Drag and drop and the more menu are usable when a block isn't selected.

I think the main thing is that this is going to be a confusing accessibility situation, since this List View contains buttons that say they will select the block, but they actually do nothing.

My opinion is that the best thing would be to not use `BlockEditorProvider` for this purpose, but find solutions within List View for this problem.

Comment (draganescu) [reply packages/block-library/src/navigation/edit/index.js]: Made a todo to give up using BlockEditorProvider 👌🏻

File: packages/block-library/src/navigation/edit/index.js (hunk 1 above)
Comment (draganescu) [reply packages/block-library/src/navigation/edit/index.js]: :D no - copy paste magic

**Comment by jameskoster** [comment]
> Perhaps a question for @jameskoster — I can't recall the latest on clickthrough. When I select a template with clickthrough, there's a blue overlay. There isn't in this branch when I select the menu. Should there be?

The overlay appears only on hover at the moment, because once the parent is selected all the child-blocks can be manipulated.

If we pursued something like https://github.com/WordPress/gutenberg/issues/43608, then the overlay would persist upon selection as well.

**Comment by jasmussen** [comment]
Good points, thanks for responding. The main issue to solve here is to ensure a great navigation building experience that lets you use the inspector interface with the main navigation block selected, to build out your entire menu. 

The tricky balance to find here, is one of not directly locking the canvas view from being edited there, but nudging towards the ergonomics of the list view. Perhaps something like the recently landed "content lock" can be employed here? ([Here are some GIFs](https://github.com/WordPress/gutenberg/pull/43037#issuecomment-1209273809)) — I know that's designed for patterns, but can that be turned on by default for navigation?

**Comment by jameskoster** [comment]
templateLock:contentOnly would theoretically allow folks to change link labels and destinations, but not re-order, add, or delete. I'm not sure that's a great fit for navigation, but I suppose it depends on which on-canvas actions we want to allow. 

Perhaps something like https://github.com/WordPress/gutenberg/issues/43608 could work. IE: 

* Selecting a menu doesn't immediately unlock any edit-ability. 
* Clicking the List button in the toolbar would toggle the visibility of the Inspector (and list view therein). 
* A separate "Edit" button (either in the toolbar or ellipsis menu) could temporarily unlock full on-canvas edit-ability.

**Comment by jasmussen** [comment]
> templateLock:contentOnly would theoretically allow folks to change link labels and destinations, but not re-order, add, or delete. I'm not sure that's a great fit for navigation, but I suppose it depends on which on-canvas actions we want to allow.

We'd still have the full "Modify" action (which btw can move to the ellipsis), and we would pair it with the inspector interface for building and adding to the menu.

**Comment by jameskoster** [comment]
I suppose we can try it, but I think it might be strange if it's possible to edit menu labels, but not add links or move things around 🤔 

A full-lock-with-easy-unlock feels like a better approach to me, similar to what we've explored for template parts.

**Comment by jasmussen** [comment]
> I suppose we can try it, but I think it might be strange if it's possible to edit menu labels, but not add links or move things around 🤔

This would work in the list view no?

I don't have a strong opinion, honestly. What I believe to be key here is: 

a) building your navigation block in the canvas needs work still
b) for the 80% of use cases, a simpler interface is likely going to be an inspector tree-view, reminiscent of the old menu editor

b mainly works if you don't lose your place every time you select different menu items in the canvas.

**Comment by draganescu** [comment]
Based on [followup in the original issue](https://github.com/WordPress/gutenberg/issues/42257#issuecomment-1276001598) this PR will move forward with attempting a customization of the list view to support opting out of canvas selection of blocks. We can merge it then just like that.

I will make it a gutenberg experiment from the start considering how many iterations  will take before the off canvas editing is really good enough UX.

**Comment by draganescu** [comment]
I'll try to move this forward in an experimental way where I copy the items I change so that iterations can land independently of trunk versions. For instance I'll see if making a copy of ListView in the nav block works and the same for other things that don't work out of the box for the requirements of off canvas editing.

**Comment by jasmussen** [comment]
What's the best way to test this? I'm not seeing any changes in the nav inspector just diving in now. 

I remain convinced that this sliding inspector editing interface for child blocks be immensely useful for the navigation block, potentially letting us reach a point of usability that reduces the pressure on all the separate rough edges the project has uncovered. Not just that, but it will also be useful for social links, buttons, content-only editing containers, and more. It has such potential, let me know any way we can help push this forward! CC: @WordPress/gutenberg-design

**Comment by draganescu** [comment]
> What's the best way to test this? I'm not seeing any changes in the nav inspector just diving in now.

@jasmussen go to Gutenberg > Experiments and check this experiment on.

> I remain convinced that this sliding inspector editing interface for child blocks be immensely useful for the navigation block,

I am too since eons ago https://github.com/WordPress/gutenberg/issues/13690#issuecomment-516321425 I've been dreaming of a structured way to manage menus :D

**Comment by jasmussen** [comment]
Took this for a spin. Nice, coming along fast. Here's a GIF showing using the inspector menu panel to rearrange a few items, select others:

navigation

Initially selecting the nav block, moving/removing items, works well.

Two challenges are: if you select an in the inspector to edit it, that item is selected in the canvas again, so you lose your place in the inspector. A Back button is one solution to that, but [the prototype](https://github.com/WordPress/gutenberg/issues/42257#issuecomment-1284797475) shows the main block still selected in sort of a "locked fashion", which remains something to consider. Is it possible to reuse some of the mechanism from that? Here's how it looks like when editing content only patterns:

content only

<details><summary>A content only pattern</summary>

```
<!-- wp:group {"templateLock":"contentOnly","align":"wide","style":{"color":{"text":"#000000","background":"#ffffff"},"border":{"radius":"10px"}}} -->
<div class="wp-block-group alignwide has-text-color has-background" style="border-radius:10px;color:#000000;background-color:#ffffff"><!-- wp:spacer {"height":"64px"} -->
<div style="height:64px" aria-hidden="true" class="wp-block-spacer"></div>
<!-- /wp:spacer -->

<!-- wp:paragraph {"align":"center","style":{"typography":{"lineHeight":".9"}},"fontSize":"small"} -->
<p class="has-text-align-center has-small-font-size" style="line-height:.9"><strong>GET IN TOUCH</strong></p>
<!-- /wp:paragraph -->

<!-- wp:heading {"textAlign":"center","style":{"typography":{"fontSize":59,"lineHeight":"1.15"}}} -->
<h2 class="has-text-align-center" id="schedule-a-visit" style="font-size:59px;line-height:1.15"><strong>Schedule a Visit</strong></h2>
<!-- /wp:heading -->

<!-- wp:buttons {"layout":{"type":"flex","justifyContent":"center","orientation":"horizontal"}} -->
<div class="wp-block-buttons"><!-- wp:button {"width":50,"style":{"color":{"background":"#000000","text":"#ffffff"},"border":{"radius":"50px"}}} -->
<div class="wp-block-button has-custom-width wp-block-button__width-50"><a class="wp-block-button__link has-text-color has-background wp-element-button" style="border-radius:50px;color:#ffffff;background-color:#000000">Contact us</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->

<!-- wp:spacer {"height":"64px"} -->
<div style="height:64px" aria-hidden="true" class="wp-block-spacer"></div>
<!-- /wp:spacer --></div>
<!-- /wp:group -->
```

</details>

Another aspect is the sliding interface, I had pictured the same mechanism used as applies to Global Styles, this one:
sliding

I recall seeing this used in the post inspector in the past, for an older color implementation (CC: @jorgefilipecosta for thoughts on that, and content lock as well). I don't know if it has to happen in this PR, but it might be necessary to get a proper back button behavior.

**Comment by jasmussen** [comment]
CC: @SaxonF in case you have time to review some of the items above ☝️

**Comment by SaxonF** [comment]
> if you select an in the inspector to edit it, that item is selected in the canvas again, so you lose your place in the inspector. A Back button is one solution to that, but https://github.com/WordPress/gutenberg/issues/42257#issuecomment-1284797475 shows the main block still selected in sort of a "locked fashion", which remains something to consider.

Adding the back button on child block and moving the edit action to an icon will help a lot. Let's see how that feels first and then decide if we need to modify selection behaviour.

File: lib/experimental/editor-settings.php [hunk 4]
```diff
@@ -83,3 +83,15 @@ function gutenberg_enable_zoomed_out_view() {
 }
 
 add_action( 'admin_init', 'gutenberg_enable_zoomed_out_view' );
+
+/**
+ * Sets a global JS variable used to trigger the availability of zoomed out view.
```
Comment (scruffian) [inline lib/experimental/editor-settings.php]:
```suggestion
 * Sets a global JS variable used to trigger the availability of the Navigation List View experiment.
```

File: lib/experiments-page.php [hunk 5]
```diff
@@ -52,6 +52,17 @@ function gutenberg_initialize_experiments_settings() {
 			'id'    => 'gutenberg-zoomed-out-view',
 		)
 	);
+	add_settings_field(
+		'gutenberg-off-canvas-navigation-editor',
+		__( 'Off canvas navigation editor ', 'gutenberg' ),
```
Comment (scruffian) [inline lib/experiments-page.php:57]: I'm not sure about this name. I prefer Navigation list view, but I'm not 100%.

File: packages/block-library/src/navigation/edit/index.js [hunk 6]
```diff
@@ -82,6 +85,9 @@ function Navigation( {
 	hasColorSettings = true,
 	customPlaceholder: CustomPlaceholder = null,
 } ) {
+	const isOffCanvasNavigationEditorEnabled =
+		window?.__experimentalEnableOffCanvasNavigationEditor === true;
```
Comment (scruffian) [inline packages/block-library/src/navigation/edit/index.js:89]: Is it OK to access window here? Just wondering if this could be accessed from contexts where `window` is not defined (like mobile apps).

File: packages/block-library/src/navigation/edit/index.js [hunk 7]
```diff
@@ -848,48 +837,103 @@ function Navigation( {
...
+							<OffCanvasEditor
+								blocks={ innerBlocks }
+								isExpanded={ true }
+								selectBlockInCanvas={ false }
+							/>
+						</PanelBody>
+					) }
+					{ ! isOffCanvasNavigationEditorEnabled && (
+						<PanelBody title={ __( 'Menu' ) }>
+							<NavigationMenuSelector
```
Comment (scruffian) [inline packages/block-library/src/navigation/edit/index.js]: It might be worth setting some of this stuff as a const, so that we don't need to duplicate it.

**Review by scruffian — COMMENTED** [review]
This is looking good. A couple of things that might be good to do:
1. Is there a way of reusing the files from the List View block that we haven't modified? It will be easier to keep track of our changes if we only include those files that are different
2. I think we can refactor some parts of the navigation block edit component to use consts, so that we don't need to duplicate so much code.

File: package.json [hunk 8]
```diff
@@ -189,7 +189,7 @@
 		"eslint-plugin-jest-dom": "4.0.2",
 		"eslint-plugin-playwright": "0.8.0",
 		"eslint-plugin-ssr-friendly": "1.0.6",
-		"eslint-plugin-testing-library": "5.7.2",
+		"eslint-plugin-testing-library": "5.9.1",
```
Comment (draganescu) [inline package.json]: Why this change in this PR?

File: packages/block-library/src/navigation/edit/index.js [hunk 9]
```diff
@@ -833,6 +833,33 @@ function Navigation( {
 		);
 	}
 
+	const navigationMenuSelectorInstance = (
```
Comment (draganescu) [inline packages/block-library/src/navigation/edit/index.js:940]: Yea, I'd rather reafactor _all_ the usages of `NavigationMenuSelector`

File: lib/experiments-page.php (hunk 5 above)
Comment (draganescu) [reply lib/experiments-page.php:57]: Who cares :) It's gonna disappear anyway when the experiment is over.

File: packages/block-library/src/navigation/edit/index.js (hunk 6 above)
Comment (draganescu) [reply packages/block-library/src/navigation/edit/index.js:89]: Yes, that is how we handle experiments. There is no other way I've seen to look into serverside set flags. Not sure what react native does about window, but the condition is `window?` (notice the [optional chaining operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining)) so if it doesn't exist `isOffCanvasNavigationEditorEnabled` is just `null`

File: packages/block-library/src/navigation/edit/index.js (hunk 7 above)
Comment (draganescu) [reply packages/block-library/src/navigation/edit/index.js]: I agree but we should do it in a separate refactoring PR targeting trunk. Here we're only exploring this UX pattern.

**Comment by draganescu** [comment]
@scruffian re:

> Is there a way of reusing the files from the List View block that we haven't modified? It will be easier to keep track of our changes if we only include those files that are different

I don't think it matters that much. The main goal of using an experiment is to move fast. The cost / trade off is duplication and the fact than when the dust settles there is work to do to reconcile things. But that is work that we can evaluate, while the exploratory phase (here) we can't evaluate because we're not sure what the end result is. There is always the help of file compare tools, making a diff should be very easy.

**Comment by jasmussen** [comment]
For whatever unknown reason, I still select the block when clicking in the inspector:

state

This is despite fixing my local env, doing a fresh npm install && npm run dev, and I even tried it in an incognito window to make sure it wasn't Chrome extensions messing with me. 

Nevertheless, a back button on the individual inspector will help. Also excited to see the "Add menu item" button making an appearance. 

Speaking of, should there be an editable label input field in the inspector, now that it's becoming a bigger part of this?

File: package.json (hunk 8 above)
Comment (scruffian) [reply package.json]: Oh sorry that was unintentional
```suggestion
		"eslint-plugin-testing-library": "5.7.2",
```

File: packages/block-library/src/navigation/edit/index.js (hunk 9 above)
Comment (scruffian) [reply packages/block-library/src/navigation/edit/index.js:940]: The other uses pass different properties, so that's a bit more complex.

**Review by scruffian — APPROVED** [review]
LGTM

File: packages/block-editor/src/components/off-canvas-editor/block.js [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+				aria-describedby={ descriptionId }
+			>
+				{ ( { ref, tabIndex, onFocus } ) => (
+					<div className="block-editor-list-view-block__contents-container">
+						<ListViewBlockContents
+							block={ block }
+							onClick={
+								selectBlockInCanvas
+									? selectEditorBlock
+									: () => {}
```
Comment (georgeh) [inline packages/block-editor/src/components/off-canvas-editor/block.js:252]: This seems to stop [the block from getting selected @jasmussen's comment](https://github.com/WordPress/gutenberg/pull/44534#issuecomment-1299816147) (and maybe fixes #45440?)
```suggestion
									: ( event ) => { event.preventDefault() }
```

Comment (draganescu) [reply packages/block-editor/src/components/off-canvas-editor/block.js:252]: Yes.

**Comment by draganescu** [comment]
Although this is approved I will close it and follow up with four PRs: to make it easier to keep track of what it was done:
1. PR to make the experiment available
2. PR to add the duplicated list view
3. PR to PR to add the new list view prop in the duplicated comp and add the duplicated list view to the navigation block to use the new prop
4. PR to move the navigation selector in the experiment
